libcrux_crypto/src/crypto.rs

```diff
@@ -0,0 +1,578 @@
+
+        let mut msg_ctx: Vec<u8> = data.to_vec();
+        let tag = libcrux::aead::encrypt(&key, &mut msg_ctx, iv, aad).map_err(|e| match e {
+            libcrux::aead::EncryptError::InvalidArgument(
+                libcrux::aead::InvalidArgumentError::UnsupportedAlgorithm,
+            ) => CryptoError::UnsupportedAeadAlgorithm,
+            _ => CryptoError::CryptoLibraryError,
+        })?;
+
+        msg_ctx.extend_from_slice(tag.as_ref());
+        Ok(msg_ctx)
+    }
+
+    fn aead_decrypt(
+        &self,
+        alg: AeadType,
+        key: &[u8],
+        ct_tag: &[u8],
+        nonce: &[u8],
+        aad: &[u8],
+    ) -> Result<Vec<u8>, CryptoError> {
+        if ct_tag.len() < 16 || nonce.len() != 12 {
+            return Err(CryptoError::InvalidLength);
+        }
+
+        let boundary = ct_tag.len() - 16;
+        let mut c = ct_tag[..boundary].to_vec();
+        let tag = &ct_tag[boundary..];
+
+        let iv = libcrux::aead::Iv::new(nonce).map_err(|_| CryptoError::InvalidLength)?;
+        let key = aead_key(alg, key)?;
+        let tag = libcrux::aead::Tag::from_slice(tag).expect("failed despite correct length");
+
+        libcrux::aead::decrypt(&key, &mut c, iv, aad, &tag).map_err(|e| match e {
+            libcrux::aead::DecryptError::InvalidArgument(
+                libcrux::aead::InvalidArgumentError::UnsupportedAlgorithm,
+            ) => CryptoError::UnsupportedAeadAlgorithm,
+            libcrux::aead::DecryptError::DecryptionFailed => CryptoError::AeadDecryptionError,
+            _ => CryptoError::CryptoLibraryError,
+        })?;
+
+        Ok(c)
+    }
+
+    fn signature_key_gen(&self, alg: SignatureScheme) -> Result<(Vec<u8>, Vec<u8>), CryptoError> {
+        let alg = sig_alg(alg)?;
+        let mut rng = self
+            .drbg
+            .lock()
+            .map_err(|_| CryptoError::CryptoLibraryError)
+            .map(GuardedRng)?;
+
+        libcrux::signature::key_gen(alg, &mut rng).map_err(|_| CryptoError::SigningError)
+    }
+
+    fn verify_signature(
+        &self,
+        alg: SignatureScheme,
+        data: &[u8],
+        pk: &[u8],
+        signature: &[u8],
+    ) -> Result<(), CryptoError> {
+        let signature = sig(alg, signature)?;
+        libcrux::signature::verify(data, &signature, pk).map_err(|_| CryptoError::InvalidSignature)
+    }
+
+    fn sign(&self, alg: SignatureScheme, data: &[u8], key: &[u8]) -> Result<Vec<u8>, CryptoError> {
+        let alg = sig_alg(alg)?;
+        let drbg = self
+            .drbg
+            .lock()
+            .map_err(|_| CryptoError::CryptoLibraryError)?;
+
+        libcrux::signature::sign(alg, data, key, &mut GuardedRng(drbg))
+            .map_err(|_| CryptoError::SigningError)
+            .map(|sig| sig.into_vec())
+    }
+
+    fn hpke_seal(
+        &self,
+        config: HpkeConfig,
+        pk_r: &[u8],
+        info: &[u8],
+        aad: &[u8],
+        ptxt: &[u8],
+    ) -> HpkeCiphertext {
```

Comment:
Would it make sense to patch the crypto provider trait to return a `Result` here? There are quite a few unwraps in this function.

Reply:
noticed that the existing provider does the same unwraps, and that it might have been better to do this in a separate PR, but now it's done